Refactor SEI delete functionality for server

PhilMFischer opened the pull request:
The delete functionality is initially based on the eclipse refactoring framework. This provides functionality to not just undo the model changes but also recreate the files on the file system. 
Unfortunately this code is based in the _ui_ and not the _core_ plugin of the refactoring framework.
Therefore i is not usable for the Server which is headless and does not use UI code. This PR handles the following points:
- Refactor the code for deleting SEIs such as both the VirSat UI and Server can make use of it.
- Move the relevant code to the project _core_ plugin and introduce callbacks to decide where to use UI based and non-UI based code.
- Implement missing jUnit tests.
- Finally provide functionality for the server to remove a SEI using the command framework.

Commit:
Refactored Remove Command to be invoked through ed - (Task #584)

- Changed code for deleting SEIs to be invoked through the
VirSatTransactionalEditingDomain, to keep correct order of first
grapping the WorkspaceLock, than grapping the transaction.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Prepared creation of non UI delete resource command - (Task #584)

- Moved code for deleting SEIs and their folders on the file system from
ui to core plugin.
- Implemented call back function to create refactoring and UI based
operation for existing UI based calls
- Adjusted all other relevant code.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Moved files to correct plugin - (Task #584)

- Accidentally moved files to test fragment instead of core plugin.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Added test case skeleton - (Task #584)

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Implemented NoUndoDelete Operation - (Task #584)

- Implemented DeleteOperation for workspace resources.
- Implemented test cases as well.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Implemented test cases for the RemoveFileStructure Command - (Task #584)

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Moved and tested method to filter SEI selection - (Task #584)

- Moved method to filter SEIs for their parent from delete code to
SEIHelper.
- Added a test case for it as well.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Added Test Code for Delete Command Creation - (Task #584)

- Added TestCases for CreateRemoveSeiWithFileStructure command factory

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Improving Command Code - (Task #584)

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Added handling for non contained resources - (Task #584)


---
Task #584: Implement non UI Command for deleting Structural Element
Instances

SaMuellerDLR commented on de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/structural/util/StructuralElementInstanceHelper.java:
```diff
@@ -108,4 +111,25 @@ public ICategoryAssignmentContainer getRoot() {
+		ArrayList<StructuralElementInstance> seisToDeleteWithoutDuplicateChildren = new ArrayList<>();
+		for (StructuralElementInstance sei : seisToDelete) {
+			EObject parent = sei.eContainer();
+			while (parent instanceof StructuralElementInstance && !selectedToBeDeleted.contains(parent)) {
```
Are we really interested in checking the type here or is it just a question of null/non-null? 
`while (parent != null && selectedToBeDeleted.contains(parent) { ... }` and `if (parent != null) { ... }` would be easier to read.

PhilMFischer replied:
No idea, I just moved and tested the already existing code. We have to investigate this.

PhilMFischer replied:
I think the way it is implemented is fine:
´´´
EObject parent = sei.eContainer();
while (parent instanceof StructuralElementInstance && !selectedToBeDeleted.contains(parent)) {
   parent = parent.eContainer();
}
´´´
parent is of type EObject. Even though our data model won't return anything else than a SEI or null, it is not practically checked when just using _null_ checks. I think _spotbugs_ may complain and we would have to introduce the type checking somewhere else. I can maybe add a comment to explain what is meant here.

PhilMFischer replied:
Looking deeper into the method, it looks like the algorithm is not correct. Therefore back to WIP.

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/command/CreateRemoveSeiWithFileStructureCommandTest.java:
```diff
@@ -0,0 +1,178 @@
+		super.setUp();
+		addEditingDomainAndRepository();
+
+		editingDomain.getCommandStack().execute(new RecordingCommand(editingDomain) {
```
Can we use the `executeAsCommand(...)` here?

PhilMFischer replied:
Absolutely! Forgot that we provided this back in the days...

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/command/CreateRemoveSeiWithFileStructureCommandTest.java:
```diff
@@ -0,0 +1,178 @@
+
+	@After
+	public void tearDown() throws CoreException {
+		super.tearDown();
```
Can be removed.

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/command/RemoveFileStructureCommandTest.java:
```diff
@@ -0,0 +1,192 @@
+		
+		addEditingDomainAndRepository();
+
+		editingDomain.getCommandStack().execute(new RecordingCommand(editingDomain) {
```
Can we use the executeAsCommand(...) here?

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/command/RemoveFileStructureCommandTest.java:
```diff
@@ -0,0 +1,192 @@
+
+	@After
+	public void tearDown() throws CoreException {
+		super.tearDown();
```
Can be removed.

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/structure/operation/NoUndoDeleteResourceOperationTest.java:
```diff
@@ -0,0 +1,132 @@
+
+	@After
+	public void tearDown() throws CoreException  {
+		super.tearDown();
```
Can be removed.

SaMuellerDLR commented on de.dlr.sc.virsat.project/src/de/dlr/sc/virsat/project/structure/command/RemoveFileStructureCommand.java:
```diff
@@ -0,0 +1,114 @@
+		executeExclusive(() -> {
+			try {
+				deleteResourceOperation.execute(null, null);
+				editingDomain.getVirSatCommandStack().triggerSaveAll();
```
Could we maybe move the triggerSaveAll() into the executeExclusive(...) method since its logically needed in all cases?

PhilMFischer replied:
Let's give it a try. When i fixed the SWTBot for stability, I realized that these changes easily ended up in deadlocks. Since we solved our deadlock issues, it is worth giving it a try. SWTBot should bail out in case things are bad.

Commit:
Minor fixes due to review - (Task #584)

- Removed some tear down.
- Shifted the trigger save all in the RemoveCommand.
- ...

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Improved comments and test for selecting SEIs for delete - (Task #584)

- Improved comments in the code.
- Improved test cases.

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

SaMuellerDLR reviewed (approved):
Approved!

PhilMFischer commented:
Closes #584
